examples/fluids/qfunctions/newtonian.h

```diff
@@ -63,6 +63,94 @@ CEED_QFUNCTION(ICsNewtonianIG_Conserv)(void *ctx, CeedInt Q, const CeedScalar *c
+  // Inputs
+  const CeedScalar(*q_data)[CEED_Q_VLA] = (const CeedScalar(*)[CEED_Q_VLA])in[0];
+  const CeedScalar(*q_true)[CEED_Q_VLA] = (const CeedScalar(*)[CEED_Q_VLA])in[1];
+  const CeedScalar(*q_soln)[CEED_Q_VLA] = (const CeedScalar(*)[CEED_Q_VLA])in[2];
+
+  // Outputs
+  CeedScalar(*q_error)[CEED_Q_VLA] = (CeedScalar(*)[CEED_Q_VLA])out[0];
+
+  // Quadrature Point Loop
+  CeedPragmaSIMD for (CeedInt i = 0; i < Q; i++) {
+    CeedScalar q_t[5] = {0}, q_s[5] = {0};
+    for (CeedInt j = 0; j < 5; j++) {
+      q_s[j] = q_soln[j][i];
+      q_t[j] = q_true[j][i];  // Automatically provided in the same variables as the solution.
+    }
+    for (CeedInt j = 0; j < 5; j++) q_error[j][i] = q_data[0][i] * (q_t[j] - q_s[j]) * (q_t[j] - q_s[j]);
+  }  // End of Quadrature Point Loop
+  return 0;
+}
+
+// *****************************************************************************
+// This QFunction computes the L2 error in primitive variables if
+//   the simulation is run in conservative variables.
+// *****************************************************************************
+CEED_QFUNCTION(Newtonian_L2ErrorAsPrimitive)(void *ctx, CeedInt Q, const CeedScalar *const *in, CeedScalar *const *out) {
+  // Inputs
+  const CeedScalar(*X)[CEED_Q_VLA]             = (const CeedScalar(*)[CEED_Q_VLA])in[0];
+  const CeedScalar(*q_data)[CEED_Q_VLA]        = (const CeedScalar(*)[CEED_Q_VLA])in[1];
+  const CeedScalar(*q_source_true)[CEED_Q_VLA] = (const CeedScalar(*)[CEED_Q_VLA])in[2];
+  const CeedScalar(*q_source_soln)[CEED_Q_VLA] = (const CeedScalar(*)[CEED_Q_VLA])in[3];
+
+  // Outputs
+  CeedScalar(*q_target_error)[CEED_Q_VLA] = (CeedScalar(*)[CEED_Q_VLA])out[0];
+
+  NewtonianIdealGasContext context = (NewtonianIdealGasContext)ctx;
+  // Quadrature Point Loop
+  CeedPragmaSIMD for (CeedInt i = 0; i < Q; i++) {
+    const CeedScalar x[3]     = {X[0][i], X[1][i], X[2][i]};
+    CeedScalar       q_c_t[5] = {0}, q_c_s[5] = {0}, q_p_t[5] = {0}, q_p_s[5] = {0};
+    for (CeedInt j = 0; j < 5; j++) {
+      q_c_s[j] = q_source_soln[j][i];
+      q_c_t[j] = q_source_true[j][i];  // Automatically provided in the same variables as the solution.
+    }
+    State s_t = StateFromU(context, q_c_t, x);
+    StateToY(context, s_t, q_p_t);
+    State s_s = StateFromU(context, q_c_s, x);
+    StateToY(context, s_s, q_p_s);
+    for (CeedInt j = 0; j < 5; j++) q_target_error[j][i] = q_data[0][i] * (q_p_t[j] - q_p_s[j]) * (q_p_t[j] - q_p_s[j]);
+  }  // End of Quadrature Point Loop
+  return 0;
+}
+
+// *****************************************************************************
+// This QFunction computes the L2 error in conservative variables if
+//   the simulation is run in primitive variables.
+// *****************************************************************************
+CEED_QFUNCTION(Newtonian_L2ErrorAsConservative)(void *ctx, CeedInt Q, const CeedScalar *const *in, CeedScalar *const *out) {
```

Comment:
We should probably combine all `Newtonian_L2ErrorAs*` into a single helper function that takes `StateVariable` as an argument and then have individual `CEED_QFUNCTION`s for the conservative and primitive case.

I also feel like we could combine the normal `Newtonian_L2Error` function into that same system; looks like the only difference is the lack state conversions at the end. ie. If the desired error and solution states are the same, then nothing would be done.

Reply:
For folding in `Newtonian_L2Error`, you could just have it's `CEED_QFUNCTION` use `context->state_var` for an if statement.

Reply:
Thanks @jrwrigh , good call. I combined `Newtonian_L2ErrorAs*` in a single QFunction and I don't think we would even need a helper function. see https://github.com/CEED/libCEED/pull/1109/commits/81a3eaea3923a79a50cbe4228df7d469c6c0a522. 

Regarding folding everything in `Newtonian_L2Error`, I would need this qfunction for a separate ceed operator in a single simulation. i.e. one operator would compute the L2 errors in the source state variable and another one would take care of the conversion. Since we would need two separate ceed operators, I was thinking that having two distinct qfunctions would make the code cleaner rather than having one qfunction that carries a flag which will be passed to the qfunction depending on which ceed operator is being called. That said, is it even possible for two ceed operators to share a qfunction?
Another reason I am not inclined to combine `Newtonian_L2ErrorAs*` and `Newtonian_L2Error` is that we don't have the primitive variable support for traveling vortex and 2D advection. We need to compute the L2 error for these problems (which I need to fix) but we can't do the conversion. Even if my first argument wasn't valid, I think we would still need to refactor euler (and possibly advection) to use the newton solver.

Reply:
It is possible for two operators to share a qfunction, but you could move the logic into one helper and have a near-empty qfunction call it with an extra parameter to select prim vs cons.

Reply:
I think [this](https://github.com/CEED/libCEED/pull/1109/commits/71306b688a28ee6f36d6d0ce689623e0c4f9b8c0) is what you mean, right? I tried this first but then didn't see any point in keeping the helper function. Could you enlighten me?

Reply:
Done! Thanks for the input.

Reply:
Looks good to me now.

Reply:
Is `Newtonian_L2Error` not just the special case of `Newtonian_ConvertL2Error` where `target_state_var == context->state_var`? 

I guess the compiler wouldn't be able to optimize for the case of them being identical (ie. remove the `StateFromQ` and `StateToQ` statements), but this isn't a performance critical chunk of code (large scale simulations aren't going to have exact solutions anyways). I'd lean towards a more compact code in this case over the small performance uplift.

Reply:
> Is `Newtonian_L2Error` not just the special case of `Newtonian_ConvertL2Error` where `target_state_var == context->state_var`?

That's true. After stepping away from my keyboard, I realized I misread your comment. For some reason I though you were talking about a single QFunction while you were talking about a single helper function for all cases. Do you like the last version better?

> I guess the compiler wouldn't be able to optimize for the case of them being identical (ie. remove the `StateFromQ` and `StateToQ` statements), but this isn't a performance critical chunk of code (large scale simulations aren't going to have exact solutions anyways). I'd lean towards a more compact code in this case over the small performance uplift.

True, I am not worried about the performance either (the logic seems a little weird though and I am not sure about its readability).

Reply:
Yep! That looks good to me!